src/data.js

```diff
@@ -263,45 +263,154 @@ let Artifact = Entity.configure({
+     *   md5: hex encoded MD5 digest of content
+     *   sha256: hex encoded SHA256 digest of content
+     * NOTE: the bucket/key terminology is s3 terminology, but other blob
+     * stores have conceptually equivalent notions.  We chose to use s3
+     * terminology because that is widely known and so far our only useful
+     * implementation of the blob storage type.  In azure, these are containers
+     * and blobs, respectively.  In google cloud storage, these are bucket and
+     * object, respectively.
      *
      * storageType: reference
      *   url:           URL that artifact should redirect to
      *
      * storageType: error
      *   reason:        Formalized reason for error artifact, see JSON schema.
      *   message:       Human readable error message to return
+     *
+     * storageType: s3 (deprecated)
+     *   bucket:        S3 bucket that contains the object
+     *   prefix:        Prefix (path) for the object within the bucket
+     *
+     * storageType: azure (deprecated)
+     *   container:     Azure container that holds the blob
+     *   path:          Path to blob within container
+     *
      */
-    details:        Entity.types.JSON,
-    expires:        Entity.types.Date,
+    details: Entity.types.JSON,
+    expires: Entity.types.Date,
+    present: Entity.types.Number, // 0 or 1 to represent whether the artifact is there
+  },
+  migrate: function (item) {
+    // Add the present flag.  We will set this value to 1 for all cases because
+    // they are either stored in this table and can be worked on atomically so
+    // are always present if the entity can be loaded or because they are the
+    // deprecated s3 or azure storageType which will never have checking for
+    // presence
+    return {
+      taskId: item.taskId,
+      runId: item.runId,
+      name: item.name,
+      storageType: item.storageType,
+      contentType: item.contentType,
+      details: item.details,
+      expires: item.expires,
+      present: 1,
+    };
   },
   context: [
-    'blobStore',      // BlobStore instance wrapping Azure Blob Storage
-    'privateBucket',  // Private artifact bucket wrapping S3
-    'publicBucket',   // Public artifact bucket wrapping S3
-    'monitor',        // base.monitor instance
+    'blobStore', // BlobStore instance wrapping Azure Blob Storage
+    'privateBucket', // Private artifact bucket wrapping S3
+    'publicBucket', // Public artifact bucket wrapping S3
+    'monitor', // base.monitor instance
+    's3', // Mapping of aws region to aws sdk object
   ],
 });
 
 /** Return JSON representation of artifact meta-data */
 Artifact.prototype.json = function() {
   return {
-    storageType:      this.storageType,
-    name:             this.name,
-    expires:          this.expires.toJSON(),
-    contentType:      this.contentType,
+    storageType: this.storageType,
+    name: this.name,
+    expires: this.expires.toJSON(),
+    contentType: this.contentType,
```

Comment:
Because i don't have whatever thing it is you do that automatically does that fancy indentation.  I use the vim auto-format stuff which just gives you nice <=80 char lines, which doesn't understand the indentation format where things line up.

I can change comments though